### PR TITLE
GNN Model for Semi-Supervised Node Classification--47973350

### PR DESCRIPTION
**Pull request by Polaris0616**
This pull request implements a GNN model for semi-supervised node classification using the Facebook Page-Page Network dataset. The following key features are included:

- Implemented a multi-layer GNN model using Graph Attention Networks  for node classification.
- Modularized the project with a `modules.py` file for the GNN model and a `dataset.py` file for the dataset loading.
- Trained the model with L2 regularization, learning rate scheduling, and early stopping for optimal performance.
- Visualized the learned node embeddings using UMAP for dimensionality reduction.
- Provided loss and accuracy tracking throughout the training process, including loss plots.
- Added test scripts to evaluate the model's performance on the test set.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 5 | Some overfitting seen. No early stop|
| Implementation functions as intended	| 3 | 3 | 2 layer GAT, torch_geometric/GATConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **15**   | |

**TSNE / UMAP:** 
Good.
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Would have been good to have test/train loss plots.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Minimal.</td><td>-1</td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Spans 2 days, 11 commits.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Test/train results and graphs missing. No references.</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. Add some info about the specific implementation here.</td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-12</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Keyi Zheng

**Comment by shakes76** [comment]
Incorrect file structure, all code needs to be in its own folder etc. -2

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.

**Comment by Polaris0616** [comment]
Dear teacher,
I have merged again, can you help me check it, I am not sure if it is right now.
Thanks!
Keyi Zheng



发自我的iPhone


------------------ Original ------------------
From: Shakes ***@***.***&gt;
Date: Tue,Nov 19,2024 10:51 AM
To: shakes76/PatternAnalysis-2024 ***@***.***&gt;
Cc: Polaris0616 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [shakes76/PatternAnalysis-2024] GNN Model for Semi-SupervisedNode Classification--47973350 (PR #69)





 
Incorrect file structure, all code needs to be in its own folder etc. -2
